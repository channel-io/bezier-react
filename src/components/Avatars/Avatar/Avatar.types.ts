/* Internal dependencies */
import { ChildrenComponentProps } from '../../../types/ComponentProps'
import InjectedInterpolation from '../../../types/InjectedInterpolation'
import { StatusType } from '../../Status'

export enum AvatarSize {
  Size20 = 20,
  Size24 = 24,
  Size30 = 30,
  Size36 = 36,
  Size42 = 42,
  Size48 = 48,
  Size90 = 90,
  Size120 = 120,
}

export default interface AvatarProps extends ChildrenComponentProps {
  avatarUrl?: string
  fallbackUrl?: string
  name: string
  size?: AvatarSize
  showBorder?: boolean
  status?: StatusType
  disabled?: boolean
  wrapperClassName?: string
  wrapperInterpolation?: InjectedInterpolation
  onClick?: (event: React.MouseEvent<HTMLDivElement>) => void
  onMouseEnter?: (event: React.MouseEvent<HTMLDivElement>) => void
  onMouseLeave?: (event: React.MouseEvent<HTMLDivElement>) => void
}
