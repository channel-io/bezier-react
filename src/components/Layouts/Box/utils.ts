/* External dependencies */
import { CSSObject } from 'styled-components'
import { isNil, isNumber, isString, omit } from 'lodash-es'

/* Internal dependencies */
import { Foundation } from 'Foundation'
import { omittedPropKeys } from './const'

/**
 * @link https://styled-components.com/docs/advanced#style-objects
 * @description Style Objects로 prop을 전달받으면서 불필요한 prop도 함께 섞여 들어옵니다. 이를 삭제합니다.
 */
export function getSanitizedCSSObject(props: CSSObject) {
  return omit(props, omittedPropKeys)
}

export function convertThemeProps(foundation?: Foundation, ...values: unknown[]) {
  for (const index in values) {
    if (Object.prototype.hasOwnProperty.call(values, index)) {
      const eachValue = values[index]
      if (!isNil(eachValue)) {
        if (isString(eachValue)) {
          return foundation?.theme?.[eachValue] ?? eachValue
        }
      }
    }
  }
  return undefined
}

export function convertSpacingProps(foundation?: Foundation, ...values: unknown[]) {
  for (const index in values) {
    if (Object.prototype.hasOwnProperty.call(values, index)) {
      const eachValue = values[index]
      if (!isNil(eachValue)) {
        if (isNumber(eachValue)) {
          return eachValue === 0 ? eachValue : `${eachValue}px`
        }
        if (isString(eachValue)) {
          return foundation?.spacing?.[eachValue] ?? eachValue
        }
      }
    }
  }
  return undefined
}
