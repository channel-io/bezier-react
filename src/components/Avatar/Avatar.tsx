/* External dependencies */
import React, { forwardRef, useCallback, useMemo } from 'react'
import _ from 'lodash'
import { v4 as uuid } from 'uuid'

/* Internal denpendencies */
import Icon from '../Icon/Icon'
import { IconSize } from '../Icon/Icon.types'
import { isLastIndex } from '../../utils/arrayUtils'
import { StyledAvatar, StyledAvatarGroup, AvatarEllipsisWrapper, AvatarEllipsis } from './Avatar.styled'
import { AvatarSize, AvatarProps, AvatarGroupProps } from './Avatar.types'

export const AVATAR_TEST_ID = 'ch-design-system-avatar'
export const AVATAR_GROUP_TEST_ID = 'ch-design-system-avatar-group'

function Avatar({
  avatarUrl,
  size = AvatarSize.M,
  name = '',
  testId = AVATAR_TEST_ID,
  disabled = false,
  showBorder = false,
  onClick = _.noop,
  onMouseEnter = _.noop,
  onMouseLeave = _.noop,
}: AvatarProps,
forwardedRef: React.Ref<HTMLDivElement>,
) {
  return (
    <StyledAvatar
      ref={forwardedRef}
      data-testid={testId}
      avatarUrl={avatarUrl}
      size={size}
      role="img"
      aria-label={name}
      showBorder={showBorder}
      disabled={disabled}
      onClick={onClick}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
    />
  )
}

export default forwardRef(Avatar)

// TODO: 올바른 페어의 ellipsis 아이콘 사이즈를 지정해줘야함
function getProperIconSize(avatarSize: AvatarSize) {
  return {
    [AvatarSize.XXS]: IconSize.XXS,
    [AvatarSize.XS]: IconSize.XS,
    [AvatarSize.S]: IconSize.S,
    [AvatarSize.M]: IconSize.Normal,
    [AvatarSize.L]: IconSize.Normal,
    [AvatarSize.XL]: IconSize.L,
    [AvatarSize.XXL]: IconSize.L,
    [AvatarSize.XXXL]: IconSize.L,
  }[avatarSize]
}

export function AvatarGroup({
  max,
  size = AvatarSize.M,
  spacing = 4,
  children,
}: AvatarGroupProps) {
  const renderAvatarElement = useCallback((avatar: React.ReactElement) => (
    React.cloneElement(avatar, {
      key: uuid(),
      size,
      showBorder: spacing < 0,
    })
  ), [
    size,
    spacing,
  ])

  const AvatarListCount = useMemo(() => (
    React.Children.count(children)
  ), [children])

  const AvatarListComponent = useMemo(() => {
    if (AvatarListCount <= max) {
      return React.Children.map(children, (avatar) => (
        React.isValidElement(avatar) && renderAvatarElement(avatar)
      ))
    }

    const sliceEndIndex = max - AvatarListCount
    const slicedAvatarList = React.Children.toArray(children).slice(0, sliceEndIndex)
    const calculatedSpacing = AvatarListCount > 0 ? spacing : 0

    return slicedAvatarList.map((avatar, index, arr) => {
      if (!React.isValidElement(avatar)) { return null }
      if (!isLastIndex(arr, index)) { return renderAvatarElement(avatar) }
      return (
        <AvatarEllipsisWrapper
          key={uuid()}
          spacing={calculatedSpacing}
        >
          <AvatarEllipsis>
            <Icon
              size={getProperIconSize(size)}
              name="more"
              color="bg-white-absolute"
            />
          </AvatarEllipsis>
          { renderAvatarElement(avatar) }
        </AvatarEllipsisWrapper>
      )
    })
  }, [
    max,
    size,
    spacing,
    children,
    AvatarListCount,
    renderAvatarElement,
  ])

  return (
    <StyledAvatarGroup
      data-testid={AVATAR_GROUP_TEST_ID}
      spacing={spacing}
    >
      { AvatarListComponent }
    </StyledAvatarGroup>
  )
}
