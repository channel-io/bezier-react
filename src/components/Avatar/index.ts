import Avatar from './Avatar'
import { AvatarSize, EmojiAvatarSize } from './Avatar.types'
import type AvatarProps from './Avatar.types'

export type {
  AvatarProps,
}

export {
  Avatar,
  AvatarSize,
  EmojiAvatarSize,
}
