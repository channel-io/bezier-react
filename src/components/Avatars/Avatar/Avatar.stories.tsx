/* External dependencies */
import React from 'react'
import { base } from 'paths.macro'
import { Story, Meta } from '@storybook/react'

/* Internal dependencies */
import { styled } from '../../../foundation'
import { getTitle } from '../../../utils/etcUtils'
import { StatusType } from '../../Status'
import Avatar from './Avatar'
import { AvatarProps, AvatarSize } from './Avatar.types'

const MOCK_AVATAR_URL = 'https://cf.channel.io/thumb/200x200/pub-file/1/606d87d059a6093594c0/ch-symbol-filled-smiley-bg.png'

const AvatarSizeList = Object.keys(AvatarSize)
  .filter(value => Number.isNaN(Number(value)) === true)
  .map(key => AvatarSize[key])

export default {
  title: getTitle(base),
  component: Avatar,
  argTypes: {
    size: {
      control: {
        type: 'radio',
        options: AvatarSizeList,
      },
    },
    status: {
      control: {
        type: 'radio',
        options: StatusType,
      },
    },
  },
} as Meta

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 200px;
  height: 200px;
  background-color: ${({ foundation }) => foundation?.theme?.['bg-grey-light']};
`

const Template: Story<AvatarProps> = (args) => (
  <Wrapper>
    <Avatar {...args} />
  </Wrapper>
)

export const Primary: Story<AvatarProps> = Template.bind({})
Primary.args = {
  avatarUrl: MOCK_AVATAR_URL,
  name: 'Channel',
  size: AvatarSize.XS,
  showStatus: true,
  status: StatusType.NONE,
  showBorder: false,
  disabled: false,
}
