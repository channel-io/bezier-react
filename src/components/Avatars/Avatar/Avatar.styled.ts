/* Internal denpendencies */
import { styled, css, smoothCorners } from '../../../foundation'
import InjectedInterpolation from '../../../types/InjectedInterpolation'
import DisabledOpacity from '../../../constants/DisabledOpacity'
import { AVATAR_STATUS_GAP, AVATAR_BORDER_WIDTH, AVATAR_BORDER_RADIUS_PERCENTAGE } from '../constants/AvatarStyle'
import { enableSmoothCorners } from '../../../worklets/EnableCSSHoudini'
import { AvatarSize } from './Avatar.types'

interface AvatarWrapperProps {
  disabled: boolean
}

interface AvatarProps {
  avatarUrl: string
  size: AvatarSize
  showBorder: boolean
  interpolation?: InjectedInterpolation
}

function calcStatusGap(showBorder: boolean) {
  return `${showBorder && enableSmoothCorners.current
    ? (AVATAR_BORDER_WIDTH * 2) - AVATAR_STATUS_GAP
    : -AVATAR_STATUS_GAP}px`
}

const disabledStyle = css`
  opacity: ${DisabledOpacity};
`

const smoothCornersFallbackBorderStyle = css`
  &::after {
    ${({ foundation }) => foundation?.border?.getBorder(AVATAR_BORDER_WIDTH, foundation?.theme?.['bg-white-absolute'])};

    position: absolute;
    top: -${AVATAR_BORDER_WIDTH}px;
    left: -${AVATAR_BORDER_WIDTH}px;
    z-index: -1;
    box-sizing: content-box;
    display: block;
    width: 100%;
    height: 100%;
    content: '';
    background-color: ${({ foundation }) => `${foundation?.theme?.['bg-white-absolute']}`};
    border-radius: ${AVATAR_BORDER_RADIUS_PERCENTAGE}%;
  }
`

export const StyledAvatar = styled.div<AvatarProps>`
  position: relative;
  box-sizing: content-box;
  display: flex;
  flex-shrink: 0;
  align-items: center;
  justify-content: center;
  width: ${({ size }) => size}px;
  height: ${({ size }) => size}px;
  outline: none;

  ${({ showBorder }) => (!enableSmoothCorners.current && showBorder) && smoothCornersFallbackBorderStyle}

  ${({ foundation, avatarUrl, showBorder }) => smoothCorners({
    shadow: showBorder ? `0 0 0 ${AVATAR_BORDER_WIDTH}px ${foundation?.theme?.['bg-white-absolute']}` : undefined,
    shadowBlur: showBorder ? AVATAR_BORDER_WIDTH : 0,
    backgroundColor: foundation?.theme?.['bg-white-absolute'],
    borderRadius: `${AVATAR_BORDER_RADIUS_PERCENTAGE}%`,
    backgroundImage: avatarUrl,
  })};

  ${({ interpolation }) => interpolation}
`

export const AvatarWrapper = styled.div<AvatarWrapperProps>`
  position: relative;
  z-index: 1;

  ${({ disabled }) => disabled && disabledStyle};
`

export const StatusWrapper = styled.div<Pick<AvatarProps, 'showBorder'>>`
  position: absolute;
  right: ${({ showBorder }) => calcStatusGap(showBorder)};
  bottom: ${({ showBorder }) => calcStatusGap(showBorder)};
`
