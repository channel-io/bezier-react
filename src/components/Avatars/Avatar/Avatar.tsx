/* External dependencies */
import React, { forwardRef, useMemo } from 'react'
import { noop, isEmpty } from 'lodash-es'

/* Internal denpendencies */
import useProgressiveImage from '../../../hooks/useProgressiveImage'
import defaultAvatarUrl from '../assets/defaultAvatar.svg'
import { Status } from '../../Status'
import AvatarProps, { AvatarSize } from './Avatar.types'
import { AvatarImage, AvatarImageWrapper, AvatarWrapper, StatusWrapper } from './Avatar.styled'

// TODO: 테스트 코드 작성
const AVATAR_WRAPPER_TEST_ID = 'bezier-react-avatar-wrapper'
export const AVATAR_TEST_ID = 'bezier-react-avatar'

function Avatar({
  avatarUrl = '',
  fallbackUrl = defaultAvatarUrl,
  size = AvatarSize.Size24,
  name,
  testId = AVATAR_TEST_ID,
  disabled = false,
  showBorder = false,
  status,
  onClick = noop,
  onMouseEnter = noop,
  onMouseLeave = noop,
  className,
  wrapperClassName,
  interpolation,
  wrapperInterpolation,
  children,
}: AvatarProps,
forwardedRef: React.Ref<HTMLDivElement>,
) {
  const loadedAvatarUrl = useProgressiveImage(avatarUrl, fallbackUrl)

  const StatusComponent = useMemo(() => {
    if (
      (isEmpty(children) && !status)
      || (children && !React.isValidElement(children))
    ) {
      return null
    }

    const Contents = (() => {
      if (children) { return children }
      if (status) {
        return (
          <Status type={status} />
        )
      }
      return null
    })()

    return (
      <StatusWrapper
        showBorder={showBorder}
      >
        { Contents }
      </StatusWrapper>
    )
  }, [
    status,
    showBorder,
    children,
  ])

  return (
    <AvatarWrapper
      className={wrapperClassName}
      interpolation={wrapperInterpolation}
      disabled={disabled}
      data-testid={AVATAR_WRAPPER_TEST_ID}
    >
      <AvatarImageWrapper>
        <AvatarImage
          className={className}
          interpolation={interpolation}
          ref={forwardedRef}
          data-testid={testId}
          avatarUrl={loadedAvatarUrl}
          size={size}
          role="img"
          aria-label={name}
          showBorder={showBorder}
          onClick={onClick}
          onMouseEnter={onMouseEnter}
          onMouseLeave={onMouseLeave}
        >
          { StatusComponent }
        </AvatarImage>
      </AvatarImageWrapper>
    </AvatarWrapper>
  )
}

export default forwardRef(Avatar)
