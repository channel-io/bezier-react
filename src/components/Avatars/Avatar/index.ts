import Avatar from './Avatar'
import type AvatarProps from './Avatar.types'
import { AvatarSize } from './Avatar.types'

export type {
  AvatarProps,
}

export {
  Avatar,
  AvatarSize,
}
