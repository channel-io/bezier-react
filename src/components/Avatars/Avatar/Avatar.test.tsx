/* External dependencies */
import React from 'react'

/* Internal dependencies */
import { render } from '../../../utils/testUtils'
import Avatar, { AVATAR_TEST_ID } from './Avatar'
import AvatarProps from './Avatar.types'

// TODO: 테스트 코드 보강
describe('Avatar test >', () => {
  let props: AvatarProps

  const mockAvatarUrl = 'https://bit.ly/dan-abramov'
  const mockFallbackUrl = 'https://www.google.com'

  beforeEach(() => {
    props = {
      avatarUrl: mockAvatarUrl,
      fallbackUrl: mockFallbackUrl,
      name: 'Name',
    }
  })

  // NOTE: unavailable smoothCorners
  const renderAvatar = (optionProps?: AvatarProps) => render(
    <Avatar {...props} {...optionProps} />,
  )

  it('Snapshot', () => {
    const { getByTestId } = renderAvatar()
    const avatar = getByTestId(AVATAR_TEST_ID)

    expect(avatar).toMatchSnapshot()
  })
})
