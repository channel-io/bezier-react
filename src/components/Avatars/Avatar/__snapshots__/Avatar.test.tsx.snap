// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar test > Snapshot 1`] = `
.c0 {
  position: relative;
  box-sizing: content-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  outline: none;
  margin: 0px;
  background-color: #FFFFFFE6;
  background-image: url(https://www.google.com);
  background-size: cover;
  border-radius: 42%;
}

@supports (background:paint(smooth-corners)) {
  .c0 {
    padding: 0px;
    margin: 0px;
    background: paint(smooth-corners);
    border-radius: 0;
    border-image-source: url(https://www.google.com);
    box-shadow: none;
    --smooth-corners: 42%;
    --smooth-corners-shadow: 0 0 0 0 transparent;
    --smooth-corners-bg-color: #FFFFFFE6;
    --smooth-corners-padding: 0;
    --smooth-corners-radius-unit: string;
  }
}

<div
  aria-label="Name"
  class="c0"
  data-testid="bezier-react-avatar"
  role="img"
  size="24"
/>
`;
